crates/shared/src/price_estimation/trade_verifier/balance_overrides.rs

```diff
@@ -71,90 +186,99 @@ impl Strategy {
+    /// the caching policy.
+    hardcoded: HashMap<Address, Strategy>,
+    /// The balance override detector and its cache. Set to `None` if
+    /// auto-detection is not enabled.
+    detector: Option<(Detector, DetectorCache)>,
 }
 
-impl Display for ConfigurationBalanceOverrides {
-    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
-        let format_entry =
-            |f: &mut Formatter, (addr, strategy): (&Address, &Strategy)| match strategy {
-                Strategy::Mapping { slot } => write!(f, "{addr:?}@{slot}"),
-            };
+impl BalanceOverrides {
+    async fn cached_detection(&self, token: Address) -> Option<Strategy> {
+        let (detector, cache) = self.detector.as_ref()?;
+        tracing::debug!(?token, "attempting to auto-detect");
 
-        let mut entries = self.0.iter();
+        {
+            let mut cache = cache.lock().unwrap();
+            if let Some(strategy) = cache.cache_get(&token) {
+                tracing::debug!(?token, "cache hit");
+                return strategy.clone();
+            }
+        }
 
-        let Some(first) = entries.next() else {
-            return Ok(());
-        };
-        format_entry(f, first)?;
+        let strategy = detector.detect(token).await;
 
-        for entry in entries {
-            f.write_str(",")?;
-            format_entry(f, entry)?;
+        // Only cache when we successfully detect the token, or we can't find
+        // it. Anything else is likely a temporary simulator (i.e. node) failure
+        // which we don't want to cache.
+        if matches!(&strategy, Ok(_) | Err(DetectionError::NotFound)) {
+            tracing::debug!(?token, ?strategy, "caching result");
+            let cached_strategy = strategy.as_ref().ok().cloned();
+            cache.lock().unwrap().cache_set(token, cached_strategy);
+        } else {
+            tracing::warn!(
+                ?token,
+                ?strategy,
+                "error auto-detecting token balance override strategy"
+            );
         }
 
-        Ok(())
+        strategy.ok()
     }
 }
 
-impl FromStr for ConfigurationBalanceOverrides {
-    type Err = anyhow::Error;
+#[async_trait::async_trait]
+impl BalanceOverriding for BalanceOverrides {
+    async fn state_override(&self, request: BalanceOverrideRequest) -> Option<StateOverride> {
+        let strategy = if let Some(strategy) = self.hardcoded.get(&request.token) {
+            tracing::debug!(token = ?request.token, "using pre-configured balance override strategy");
+            Some(strategy.clone())
+        } else {
+            self.cached_detection(request.token).await
+        }?;
 
-    fn from_str(s: &str) -> Result<Self, Self::Err> {
-        if s.is_empty() {
-            return Ok(Self::default());
-        }
+        let (key, value) = strategy.state_override(&request.holder, &request.amount);
+        tracing::debug!(?strategy, ?key, ?value, "overriding token balance");
```

Comment:
Also too noisy. Maybe it could use `trace` level so we can enable it when we actually see problems with the feature.

Reply:
I will change my `debug` -> `trace` for the logs. This way, you can see the logs in case you need to debug something (like the feature is working as expected).